### ISSUE TITLE
Suggestion: Configurable @npm module compilation before source code compile/bundle?

### ISSUE THREAD

**Issue by LStAmour** [bazelbuild/rules_nodejs#499]
# 🚀 feature request

### Relevant Rules

yarn_install / rollup_bundle / future babel rules? / ibazel?

### Description

So I'm adjusting to living without Webpack, using tsc or these days Babel, via Rollup, to convert a small es2018 React app into a minified version. But the second I add dependencies, the time increases, and if I'm creating one bundle from all my source code, a change in one place means Rollup has to process every file's changes and it can take 30 seconds for a small project with less than a dozen files, and maybe two dozen unique import lines for third-party modules. And if I'm using PostCSS/SCSS modules, Node/CommonJS resolution, and minification and babel plugins, well every little feature and custom configuration added via Rollup takes time to load, process, and further delays the bundle as a result. 

I'd have skipped the CommonJS resolution, but not everybody publishes their dist builds as native es modules, and then Google's official Material react components use Webpack syntax unsupported by CommonJS because they use __esModule magic and apparently both Closure Compiler and Rollup have decided that's evil and so you have to write ugly "namedExports" syntax to avoid error messages from Rollup. (A suggested replacement is to run a Babel code transform to convert __esModule/Webpack magic to standard es module or standard CommonJS module syntax instead... sadly such a beast does not exist, likely due to Webpack's current popularity...) Then you've modules like json2csv that natively use `require('os')` to borrow `os.EOL` without checking if `require` is a function first, let alone if that import succeeds. So they get around that by publishing a `package.json` file with a `browser` line that some module loaders can be configured to prefer over the standard nodejs main import reference. Except by default, Rollbar's CommonJS scans the file system for cjs modules and wouldn't you know it, the cjs module they publish is node-compatible, not browser-compatible. So how does a particular third-party library depending on json2csv solve this? By using yet another transform to remove NodeJS code from the imported library. Okay...

All of this extra load time, all of this extra compilation -- all the dependency resolution by Babel/Rollbar/etc. has me wondering -- is there not a better way forward, perhaps something we can encourage with Bazel's fine-grained NPM module handling?

### Describe the solution you'd like

So here's my idea -- still very rough around the edges. So the reason we have ES module syntax, CommonJS, and even Web Package formats, is because we want to take JavaScript written in a modular syntax and export it in a bundle that can be read back as separate modules. There are two constraints to the import process -- first, the code to import has to be in an understandable syntax, usually bog-standard ES5 JavaScript -- and second the module system has to be supported by the importer (so no runtime-IIFE exporting unless we're using V8 to evaluate, I suppose).

Now, the Webpack way of optimizing this, from a very minimal reading of third-party descriptions of it, which `rollbar watch` also does, is to compile everything and cache it in memory, then skip the work you've already done when responding to changed files. My thinking is -- we make this more explicit in the build process. What if, as the next logical step to using an npm module as a dependency in your code, you had to compile a version of the code optimized for your build process and import resolution? Ideally one which targets the highest version of ES/TS/Flow you support that's also supported by the third-party source code. I presume if Google code relies on third-party JS, that you'd have to run a script of some kind to convert the presumably CommonJS into goog.provide and goog.require, but I've no idea if that's actually the case. If so, though, it's the manual equivalent to what I'm suggesting. I suppose a similar process is the precompiled-JARs for Java, and how IDEs can load src JARs if you want additional benefits only source code can provide...

Can it be automated? I would think it can, with enough work or study of how Webpack and friends do it -- as today's market share leader -- but ideally you'd have configuration options as well so detail-oriented folks could pick whether they wanted to use a bundled dist build, run the third-party's build process themselves and use the output (if they want to target a different module loader or target syntax), or run a custom build process (perhaps involving Babel) to tweak the third-party code as desired and build the result using a custom build tool and a pre-configured, expected output format (perhaps with validation?). The idea is that Bazel could oversee the compilation for each module in its sandboxing and then help insert the reference to the newly compiled import, either by using Babel-style code rewriting of your sources, by changing module resolution in your build steps, or by changing what code lies within node_modules at compile-time.

Now, I can see some issues, though some we would have to live with. Because of how dependencies work in Node, if you import a dist build or run the third-party's build process, you'll have to build against the module versions they specify and not necessarily the versions you're using elsewhere. If you use the same code in two places, you could end up with duplication ultimately only resolved by gzip, careful inspection of output module sizes/contents and changing your dependencies or careful yarn configuration. (Having tried that last one, to run Yarn in single-version dependency mode, however, I know that's not an option unless Babel is involved -- newer versions of things inevitably break code and would require code transformations or pull-requests to update the APIs involved... and the alternative is your third-party dependency breaks until you fork and update it, so Yarn settings aren't a solution, merely a test case that could easily fail...) The solution to me then, for the above issue, is to try and encourage use of Closure Compiler to eliminate as much dead code as possible, and if possible, to combine identical functions across modules. And to publicize somehow the commands you can run to get Yarn to output when multiple versions of a module are required by your dependencies.

A trickier problem is what to do if a module at runtime expects node_modules code to behave a certain way... but I suppose the answer to that is to either use import statement rewriting or module resolution hacks to preserve the existing node_modules folder layout, or to make this caching/compilation step completely optional, so it's up to the Bazel build author to notice these possible issues in testing the build process or in choosing not to use a precompiled dist provided by a Node module, as there are plenty of variables with npm packaging that doesn't make this any easier, especially when SCSS and non-JS files are made available for import/use in a node module.

### Describe alternatives you've considered

The only alternative I can think of is to better integrate ibazel into rules_nodejs, encouraging the use of "babel watch", "rollup watch" or the use of webpack dev server, to serve incrementally compiled assets almost as quick as they're saved. The trouble is that if Bazel doesn't completely manage the configuration of Babel, Rollup or Webpack Dev Server entirely (including IDE plugin suggestions/configuration/linting), you'll end up with a dev/IDE process that completely diverges from the testing/build process you run via Bazel on a remote CI server, for example. Or you end up with folks suggesting that the solution is to run these commands in Docker locally and remotely, replacing chunks of Bazel with Docker and calling it a day. (I'll admit, it's occurred to me many times over the past weeks trying to use Bazel that -- Docker is just plain easier to explain to folks. It's true, it's a lot more work in that each Bazel rule would potentially get its own container, and you still have to figure out how to distribute and track source code and rule dependency graphs, but at least then I wouldn't be asking myself how to ensure source code is mounted in the same folder on my local machine and in the same path on my remote machine -- I could just run Docker in both places, with the right volume paths set up, the right system time, and a very minimal system image, and be done with it using standard tools...)

Related: #455 (a flag for how Rollup works) and the lack of third-party plugin support in the built-in rollup_bundle rule (so I wrote my own `genrule` for Rollup as a workaround, but it's not at all clean and "Bazel-like")

**Comment by LStAmour** [comment]
Ah ... upon review, I wrote far too much text here and I need to edit it down. I'll blame the late hour I'm typing this at, and will try to review it for length and brevity in the next couple days if I can find the time. I'd also want to point out that we would likely need more than one toolchain running -- it's possible you'd have to target one module syntax for JavaScript and another module syntax and export for CSS, flat files, etc. (Unless you're one of those people base64ing or CSS-in-JSing and including everything in JS modules, of course...)

**Comment by alexeagle** [comment]
I don't really understand still. Are you suggesting that we do some code generation, following the import graph? Distribute the bundler computation around the graph? Maybe an issue isn't the right format for this discussion.
